lib/web/app/auth.go

```diff
@@ -0,0 +1,119 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package app
+
+import (
+	"crypto/subtle"
+	"net/http"
+
+	"github.com/gravitational/trace"
+	"github.com/julienschmidt/httprouter"
+
+	"github.com/gravitational/teleport/api/types"
+	apievents "github.com/gravitational/teleport/api/types/events"
+	"github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/teleport/lib/httplib"
+)
+
+// handleAuth handles authentication for an app
+// When a `POST` request comes in from a trusted proxy address, it'll set the value from the
+// `X-Cookie-Value` header to the `__Host-grv_app_session` cookie.
+func (h *Handler) handleAuth(w http.ResponseWriter, r *http.Request, p httprouter.Params) error {
+	httplib.SetNoCacheHeaders(w.Header())
+
+	cookieValue := r.Header.Get("X-Cookie-Value")
+	if cookieValue == "" {
+		return trace.AccessDenied("access denied")
+	}
+
+	subjectCookieValue := r.Header.Get("X-Subject-Cookie-Value")
+	if cookieValue == "" {
+		return trace.BadParameter("X-Subject-Cookie-Value header missing")
+	}
+
+	// Validate that the caller is asking for a session that exists.
+	ws, err := h.c.AccessPoint.GetAppSession(r.Context(), types.GetAppSessionRequest{
+		SessionID: cookieValue,
+	})
+	if err != nil {
+		h.log.WithError(err).Warn("Request failed: unable to get app session")
+		return trace.AccessDenied("access denied")
+	}
+
+	if err := checkSubjectToken(subjectCookieValue, ws); err != nil {
+		h.log.Warnf("Request failed: %v.", err)
+
+		h.c.AuthClient.EmitAuditEvent(h.closeContext, &apievents.AuthAttempt{
+			Metadata: apievents.Metadata{
+				Type: events.AuthAttemptEvent,
+				Code: events.AuthAttemptFailureCode,
+			},
+			UserMetadata: apievents.UserMetadata{
+				Login: ws.GetUser(),
+				User:  "unknown",
+			},
+			ConnectionMetadata: apievents.ConnectionMetadata{
+				LocalAddr:  r.Host,
+				RemoteAddr: r.RemoteAddr,
+			},
+			Status: apievents.Status{
+				Success: false,
+				Error:   err.Error(),
+			},
+		})
+
+		return trace.AccessDenied("access denied")
+	}
+
+	// Set the "Set-Cookie" header on the response.
+	// Set Same-Site policy for the session cookie to None in order to
+	// support redirects that identity providers do during SSO auth.
+	// Otherwise the session cookie won't be sent and the user will
+	// get redirected to the application launcher.
+	//
+	// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie/SameSite
+
+	http.SetCookie(w, &http.Cookie{
+		Name:     CookieName,
+		Value:    cookieValue,
+		Path:     "/",
+		HttpOnly: true,
+		Secure:   true,
+		SameSite: http.SameSiteNoneMode,
```

Comment:
This can be made "Lax" to harden a bit the application. Just verified dynamically that works.

Reply:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie/SameSite#lax
> Cookies are not sent on normal cross-site subrequests (for example to load images or frames into a third party site), but are sent when a user is navigating to the origin site (i.e., when following a link).